### PR TITLE
Implement early bail-out

### LINKED ISSUE #5375: S2222 performance: Improve early bailout logic
Improve `LocksReleasedAllPathsBase.ShouldExecute()` logic to execute the engine only if there exist related pair of acquire&release methods.

We should stop exploring when we meet any lambda or nested function.

This needs to be done for C# and also VB.NET

### PR DESCRIPTION
**Pull request by costin-zaharia-sonarsource**
Fixes #5375

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/LocksReleasedAllPaths.cs [hunk 1]
```diff
@@ -31,8 +33,39 @@ public class LocksReleasedAllPaths : LocksReleasedAllPathsBase
...
+        }
+
+        private sealed class LockAcquireReleaseWalker : CSharpSyntaxWalker
+        {
+            private readonly LockAcquireReleaseCollector collector;
+
+            public LockAcquireReleaseWalker(LockAcquireReleaseCollector collector)
+            {
+                this.collector = collector;
+            }
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/LocksReleasedAllPaths.cs:49]: Can be arrow method

File: analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/LocksReleasedAllPaths.cs [hunk 2]
```diff
@@ -31,8 +33,39 @@ public class LocksReleasedAllPaths : LocksReleasedAllPathsBase
...
+
+            public override void Visit(SyntaxNode node)
+            {
+                if (collector.LockAcquiredAndReleased
+                    // Lambda expressions and local functions are analyzed separately.
+                    || node is LambdaExpressionSyntax or ParenthesizedLambdaExpressionSyntax
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/LocksReleasedAllPaths.cs:55]: We need to cover all cases from here:
https://github.com/SonarSource/sonar-dotnet/blob/88af512b7bba6f3095c1d42f4b39d2b422d2799d/analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs#L135-L136

```suggestion
                    || node is LambdaExpressionSyntax or ParenthesizedLambdaExpressionSyntax or AnonymousMethodExpressionSyntax
```
Test case:
```
                System.VoidDelegate anonymousMethod = delegate     
                {
                    // Something
                };
```

File: analyzers/src/SonarAnalyzer.VisualBasic/SymbolicExecution/Roslyn/LocksReleasedAllPaths.cs [hunk 3]
```diff
@@ -31,9 +31,37 @@ public class LocksReleasedAllPaths : LocksReleasedAllPathsBase
...
 
+        private sealed class LockAcquireReleaseWalker : VisualBasicSyntaxWalker
+        {
+            private readonly LockAcquireReleaseCollector collector;
+
+            public LockAcquireReleaseWalker(LockAcquireReleaseCollector collector)
+            {
+                this.collector = collector;
+            }
+
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.VisualBasic/SymbolicExecution/Roslyn/LocksReleasedAllPaths.cs:48]: Can be arrow method

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/LocksReleasedAllPathsBase.cs [hunk 4]
```diff
@@ -156,5 +162,52 @@ private ProgramState RemoveLock(SymbolicContext context, ISymbol symbol)
...
+
+            private bool lockAcquired;
+            private bool lockReleased;
+
+            public bool LockAcquiredAndReleased =>
+                lockAcquired && lockReleased;
+
+            public void RegisterIdentifier(string name)
+            {
+                lockAcquired |= IsLockType(name) || IsLockMethod(name);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/LocksReleasedAllPathsBase.cs:199]: This is bitwise operator, so it evaluates `IsLockType` and `IsLockMethod` even when we already know that `lockAcquired` is true. 
```suggestion
                lockAcquired = lockAcquired || IsLockType(name) || IsLockMethod(name);
```

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/LocksReleasedAllPathsBase.cs [hunk 5]
```diff
@@ -156,5 +162,52 @@ private ProgramState RemoveLock(SymbolicContext context, ISymbol symbol)
...
+            }
+
+            private static bool IsLockType(string name) =>
+                LockTypes.Contains(name);
+
+            private static bool IsLockMethod(string name) =>
+                LockMethods.Contains(name);
+
+            private static bool IsReleaseMethod(string name) =>
+                ReleaseMethods.Contains(name);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/LocksReleasedAllPathsBase.cs:210]: I would not hide these behind methods. The variable names are good so it read the same way without the method. And we already have a full list of methods that we need to support in this rule, so nothing dramatic will change.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.Monitor.Conditions.cs [hunk 6]
```diff
@@ -326,5 +326,30 @@ public int MyProperty
                 }
             }
         }
+
+        public void Lambda(bool condition)
+        {
+            Action action = () =>
+            {
+                Monitor.Enter(obj); // // FN, lambdas are not yet supported
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.Monitor.Conditions.cs:334]: I do not agree, it should work for both lambdas (parenthesized and nonparenthesized) and anonymous functions, see this:
https://github.com/SonarSource/sonar-dotnet/blob/88af512b7bba6f3095c1d42f4b39d2b422d2799d/analyzers/tests/SonarAnalyzer.UnitTest/Rules/SymbolicExecution/SymbolicExecutionRunnerTest.cs#L214-L235

Please add those two missing syntaxes as well, as these walkers will be the inspiration for other rules as well in the future.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/LocksReleasedAllPathsBase.cs [hunk 7]
```diff
@@ -113,8 +115,12 @@ public override void ExecutionCompleted()
             }
         }
 
-        protected static bool ShouldExecuteFor(SyntaxToken identifier) =>
-            identifier.Text.Contains("Exit") || identifier.Text.Contains("ReleaseMutex") || identifier.Text.Contains("ReleaseReaderLock");
+        public override bool ShouldExecute()
+        {
+            var collector = new LockAcquireReleaseCollector();
+            Visit(NodeContext.Node, collector);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/LocksReleasedAllPathsBase.cs:121]: There's a bug in the way it is used. If we call it with lambda entry point, the first thing the walker does is that it stops walking the tree. We can avoid it by invoking like this
```
            foreach (var child in NodeContext.Node.ChildNodes())
            {
                Visit(NodeContext.Node, collector);
            }
```
but then it doesn't make sense to recreate the walker instance inside :/

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/LocksReleasedAllPaths.cs (hunk 2 above)
Comment (costin-zaharia-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/LocksReleasedAllPaths.cs:55]: Good point. We can cover all of them using the base class `AnonymousFunctionExpressionSyntax`.

**Pull request merged**